File: src/graph/relationship_data.ts
import Data from "./data";

export type RelationshipRecord = { left_id: string; right_id: string } & Record<string, any>;

class RelationshipData extends Data {
    constructor(records: RelationshipRecord[] = []) {
        super(records);
        super._buildIndex("left_id");
        super._buildIndex("right_id");
    }
    public find(id: string, hop: number = 0, direction: "left" | "right" = "right"): boolean {
        return super._find(id, hop, direction === "left" ? "right_id" : "left_id");
    }
    /*
    ** Get the properties of the current relationship record
    '' excluding the left_id and right_id fields
    */
    public properties(): Record<string, any> | null {
        const current = this.current();
        if (current) {
            const { left_id, right_id, _type, ...props } = current;
            return props;
        }
        return null;
    }
    public current(hop: number = 0): RelationshipRecord | null {
        return super.current(hop) as RelationshipRecord | null;
    }
}

export default RelationshipData;

File: flowquery-py/src/graph/relationship_data.py
"""Relationship data class for FlowQuery."""

from typing import Any, Dict, List, Optional, TypedDict

from .data import Data


class RelationshipRecord(TypedDict, total=False):
    """Represents a relationship record from the database."""
    left_id: str
    right_id: str


class RelationshipData(Data):
    """Relationship data class extending Data with left_id and right_id indexing."""

    def __init__(self, records: Optional[List[Dict[str, Any]]] = None):
        super().__init__(records)
        self._build_index("left_id")
        self._build_index("right_id")

    def find(self, id: str, hop: int = 0, direction: str = "right") -> bool:
        """Find a relationship by node ID and direction."""
        key = "right_id" if direction == "left" else "left_id"
        return self._find(id, hop, key)

    def properties(self) -> Optional[Dict[str, Any]]:
        """Get properties of current relationship, excluding left_id, right_id, and _type."""
        current = self.current()
        if current:
            props = dict(current)
            props.pop("left_id", None)
            props.pop("right_id", None)
            props.pop("_type", None)
            return props
        return None
